### PR TITLE
Revamp constructors

### LINKED ISSUE #273: Non-sorting constructor
As defined in https://github.com/vbarrielle/sprs/issues/121#issuecomment-764796038 and the subsequent discussion.

### PR DESCRIPTION
**Pull request by mulimoen**
This moves and renames the constructors of `CsMatBase`. This makes construction of matrices for non-`Copy` types easier, and unifies constructors.

Modified:
* `new`: Now implemented on `CsMatBase` instead of `CsMatI`. Does not sort indices. Panics on failure
* `new_csc`:  Counterpart to above
* `try_new`: Returns a tuple with all owned data + error on failure
* `try_new_csc`: Counterpart to above

New:
* `new_unchecked`: Unsafe constructor for `CsMatBase`, does not do any data checks.

Removed:
* `new_view`: `new` can now be used instead
* `new_view_raw`: This function is really unsafe, we are opening ourselves to a use-after-free by making a lifetime from nothing. I'd rather push this responsibility away from this crate.

Maybe we could remove `new_vecview_raw`? This is not in use by any of our crates, and suffers from the same problems as `new_view_raw`.

Fixes #273

### CODE REVIEW

**Comment by mulimoen** [comment]
On second thought I think it would be ok to have both `new` and `try_new`. I will however change `try_new` so it returns the owned data.

File: src/sparse/csmat.rs [hunk 1]
```diff
@@ -249,6 +328,42 @@ where
             }),
         }
     }
+
+    /// Try create an owned `CSR` matrix from moved data.
```
Comment (vbarrielle) [inline src/sparse/csmat.rs:332]: Actually it's also possible to create a `CsMatViewMut` using this constructor I think.

File: src/sparse/smmp.rs [hunk 2]
```diff
@@ -439,8 +439,7 @@ mod test {
 
         let mut c_data = vec![0.; c_indices.len()];
         let mut tmp = [0.; 5];
-        let mut c = crate::CsMatViewMutI::new_trusted_mut_view(
-            crate::CSR,
+        let mut c = crate::CsMatViewMutI::new(
```
Comment (vbarrielle) [inline src/sparse/smmp.rs:442]: I'm really glad this change manages to make this work, I didn't like the call to `new_trusted_mut_view` I had to introduce before, it felt unidiomatic.

Comment (vbarrielle) [reply src/sparse/smmp.rs:442]: So now I'm wondering if that means we can remove `new_trusted_mut_view` ?

**Review by vbarrielle — APPROVED** [review]
This looks very good, I really like how it enables a more coherent API. As mentionned in the code comments, I'm wondering if it's an opportunity to remove `new_trusted_mut_view` that I introduced in the previous PR, we probably don't want to stabilize this call if it can be expressed with `new` since this refactor.

If it can be removed let's remove it before merging this PR, otherwise I'll merge it as is.

File: src/sparse/smmp.rs (hunk 2 above)
Comment (mulimoen) [reply src/sparse/smmp.rs:442]: Yes, that worked, Moved `new_trusted` to the appropriate traits restrictions and this worked

Comment (mulimoen) [reply src/sparse/smmp.rs:442]: That was a typo it seems, it should call `new_trusted` so we don't check our known valid matrice. Fixed

File: src/sparse/csmat.rs (hunk 1 above)
Comment (mulimoen) [reply src/sparse/csmat.rs:370]: I've removed all mentions of `owned` and made some examples on `new` which shows how one can construct different type (aliases) directly

File: src/sparse/smmp.rs (hunk 2 above)
Comment (vbarrielle) [reply src/sparse/smmp.rs]: Good catch, I had not noticed this would check the invariants unnecessarily.

**Review by vbarrielle — APPROVED** [review]
It's perfect now, thanks!

**Comment by mulimoen** [comment]
@vbarrielle What is your though on `new_vecview_raw`? This can easily be replaced by `new_unchecked` as was done with `sprs-ldl`

**Comment by vbarrielle** [comment]
> @vbarrielle What is your though on `new_vecview_raw`? This can easily be replaced by `new_unchecked` as was done with `sprs-ldl`

I think the more we can use `new_unchecked` or `new_trusted` to replace these calls, the better. If we can have a very coherent API it'll be easier to code against `sprs`.

**Pull request merged**